src/Esprima/Scanner.cs

```diff
@@ -1084,6 +1051,27 @@ public bool IsImplicitOctalLiteral()
             return true;
         }
 
+        private string ScanLiteralPart(Func<char, bool> check)
+        {
+            string num = "";
+
+            if (Source.CharCodeAt(Index) == '_')

```

Comment:
I think the `Source.CharCodeAt(Index)` value should be saved locally to prevent multiple calls.